### PR TITLE
Different type segments in a single PathSegmentRecords

### PR DESCRIPTION
**Pull request by pszal**
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/545%23issuecomment-160564657%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23issuecomment-160589745%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46122359%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46122477%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46122525%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46124722%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46124888%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46124895%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46125084%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46126247%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46126255%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23issuecomment-160564657%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40kormat%20%22%2C%20%22created_at%22%3A%20%222015-11-30T09%3A11%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222015-11-30T10%3A28%3A09Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2016aaf1a20a8296180490454a33c3f123d1c825b4%20lib/packet/pcb.py%204%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46125084%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%60%23%20pragma%3A%20no%20cover%60%22%2C%20%22created_at%22%3A%20%222015-11-30T10%3A05%3A34Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22done%22%2C%20%22created_at%22%3A%20%222015-11-30T10%3A18%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/pcb.py%3AL118-127%22%7D%2C%20%22Pull%20fd6717a214333e4fae7d5a6d03e4234312002d4f%20lib/packet/path_mgmt.py%20107%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46122359%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Also%20need%20to%20cover%20PST.DOWN%20and%20PST.CORE%22%2C%20%22created_at%22%3A%20%222015-11-30T09%3A38%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22and%20PST.UP_DOWN%2C%20until%20we%20get%20rid%20of%20that.%22%2C%20%22created_at%22%3A%20%222015-11-30T09%3A40%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-11-30T10%3A03%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/path_mgmt.py%3AL123-174%22%7D%2C%20%22Pull%20fd6717a214333e4fae7d5a6d03e4234312002d4f%20test/lib_packet_path_mgmt_test.py%2088%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46124722%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Ok%2C%20this%20is%20completely%20baffling.%20I%27ve%20no%20idea%20where%20those%20numbers%20come%20from%2C%20but%20they%20do%20add%20up%20to%20the%20correct%20amount.%20A%20clearer%20way%20%28at%20least%20for%20me%29%20is%3A%20%602%2B3%2B4%2B5%2B6%2B7%60%22%2C%20%22created_at%22%3A%20%222015-11-30T10%3A01%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22done%22%2C%20%22created_at%22%3A%20%222015-11-30T10%3A17%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20test/lib_packet_path_mgmt_test.py%3AL176-185%22%7D%2C%20%22Pull%20fd6717a214333e4fae7d5a6d03e4234312002d4f%20endhost/sciond.py%2025%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/545%23discussion_r46122477%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Slightly%20cleaner%20to%20do%3A%5Cr%5Cn%60%60%60%5Cr%5Cnfor%20seg_type%2C%20pcbs%20in%20path_reply.pcbs.items%28%29%3A%5Cr%5Cn%20%20for%20pcb%20in%20pcbs%3A%5Cr%5Cn%60%60%60%22%2C%20%22created_at%22%3A%20%222015-11-30T09%3A39%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-11-30T10%3A03%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/sciond.py%3AL131-164%22%7D%7D%2C%20%22approved%22%3A%20%7B%22https%3A//github.com/kormat%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/kormat'></a>
- [ ] <a href='#crh-comment-Pull fd6717a214333e4fae7d5a6d03e4234312002d4f test/lib_packet_path_mgmt_test.py 88'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/545#discussion_r46124722'>File: test/lib_packet_path_mgmt_test.py:L176-185</a></b>
- <a href='https://github.com/kormat'></a> Ok, this is completely baffling. I've no idea where those numbers come from, but they do add up to the correct amount. A clearer way (at least for me) is: `2+3+4+5+6+7`
- <a href='https://github.com/pszalach'></a> done
- [ ] <a href='#crh-comment-Pull 16aaf1a20a8296180490454a33c3f123d1c825b4 lib/packet/pcb.py 4'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/545#discussion_r46125084'>File: lib/packet/pcb.py:L118-127</a></b>
- <a href='https://github.com/kormat'></a> `# pragma: no cover`
- <a href='https://github.com/pszalach'></a> done
- [x] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/545#issuecomment-160564657'>General Comment</a></b>
- <a href='https://github.com/pszalach'></a> @kormat
- [x] <a href='#crh-comment-Pull fd6717a214333e4fae7d5a6d03e4234312002d4f lib/packet/path_mgmt.py 107'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/545#discussion_r46122359'>File: lib/packet/path_mgmt.py:L123-174</a></b>
- <a href='https://github.com/kormat'></a> Also need to cover PST.DOWN and PST.CORE
- <a href='https://github.com/kormat'></a> and PST.UP_DOWN, until we get rid of that.
- [x] <a href='#crh-comment-Pull fd6717a214333e4fae7d5a6d03e4234312002d4f endhost/sciond.py 25'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/545#discussion_r46122477'>File: endhost/sciond.py:L131-164</a></b>
- <a href='https://github.com/kormat'></a> Slightly cleaner to do:

```
for seg_type, pcbs in path_reply.pcbs.items():
for pcb in pcbs:
```

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/545?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/545?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/545?approve=0'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/545'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: lib/packet/path_mgmt.py [hunk 1]
```diff
@@ -122,50 +123,51 @@ def __init__(self, raw=None):  # pragma: no cover
...
     def __str__(self):
         s = []
         s.append("%s(%dB):" % (self.NAME, len(self)))
-        s.append("  %s" % self.info)
-        for pcb in self.pcbs:
-            s.append("  %s" % pcb)
+        if self.pcbs[PST.UP]:
+            s.append("  %s:" % PST.to_str(PST.UP))
+            for pcb in self.pcbs[PST.UP]:
+                s.append("    %s" % pcb.short_desc())
```
Comment (kormat) [inline lib/packet/path_mgmt.py]: Also need to cover PST.DOWN and PST.CORE

File: endhost/sciond.py [hunk 2]
```diff
@@ -131,26 +131,33 @@ def handle_path_reply(self, pkt):
...
-                self._handle_core_seg(pcb, first, last)
-            else:
-                logging.warning(
-                    "Incorrect path in Path Record. Info: %s PCB: %s",
-                    info.short_desc(), pcb.short_desc())
-        key = (info.seg_type, info.src_isd, info.src_ad, info.dst_isd,
-               info.dst_ad)
-        self.requests.put((key, None))
+        for seg_type in path_reply.pcbs:
+            for pcb in path_reply.pcbs[seg_type]:
```
Comment (kormat) [inline endhost/sciond.py]: Slightly cleaner to do:

```
for seg_type, pcbs in path_reply.pcbs.items():
  for pcb in pcbs:
```

File: lib/packet/path_mgmt.py (hunk 1 above)
Comment (kormat) [inline lib/packet/path_mgmt.py]: and PST.UP_DOWN, until we get rid of that.

File: test/lib_packet_path_mgmt_test.py [hunk 3]
```diff
@@ -182,10 +176,9 @@ class TestPathSegmentRecordsLen(object):
     """
     def test(self):
         inst = PathSegmentRecords()
-        inst.info = range(5)
-        inst.pcbs = [range(5) for x in range(5)]
+        inst.pcbs = {1: ['1', '22', '333', '4444'], 2: ['55555'], 3: ['666666']}
         # Call
-        ntools.eq_(len(inst), 30)
+        ntools.eq_(len(inst), 4 + 10 + 1 + 5 + 1 + 6)
```
Comment (kormat) [inline test/lib_packet_path_mgmt_test.py]: Ok, this is completely baffling. I've no idea where those numbers come from, but they do add up to the correct amount. A clearer way (at least for me) is: `2+3+4+5+6+7`

File: lib/packet/path_mgmt.py (hunk 1 above)
Comment (kormat) [inline lib/packet/path_mgmt.py]: :+1:

File: endhost/sciond.py (hunk 2 above)
Comment (kormat) [inline endhost/sciond.py]: :+1:

File: lib/packet/pcb.py [hunk 4]
```diff
@@ -118,6 +118,9 @@ def from_values(cls, isd_id, ad_id, hof, ig_rev_token=None):
         inst.ig_rev_token = ig_rev_token or bytes(REV_TOKEN_LEN)
         return inst
 
+    def get_isd_ad(self):
```
Comment (kormat) [inline lib/packet/pcb.py]: `# pragma: no cover`

File: test/lib_packet_path_mgmt_test.py (hunk 3 above)
Comment (pszal) [inline test/lib_packet_path_mgmt_test.py]: done

File: lib/packet/pcb.py (hunk 4 above)
Comment (pszal) [inline lib/packet/pcb.py]: done

**Pull request merged**